tests/components/broadlink/test_device.py

```diff
@@ -188,9 +204,58 @@ async def test_device_unload_works(hass):
         await hass.config_entries.async_unload(mock_entry.entry_id)
 
     assert mock_entry.state == ENTRY_STATE_NOT_LOADED
-    assert len(mock_forward.mock_calls) == 3
     forward_entries = {c[1][1] for c in mock_forward.mock_calls}
-    assert forward_entries == {"remote", "sensor", "switch"}
+    domains = {domain for domain, types in DOMAINS_AND_TYPES if mock_api.type in types}
+    assert mock_forward.call_count == len(domains)
+    assert forward_entries == domains
+
+
+async def test_device_unload_authentication_error(hass):
+    """Test we unload a device that failed the authentication step."""
+    device = pick_device(1)
+    mock_api = device.get_mock_api()
+    mock_api.auth.side_effect = blke.AuthenticationError()
+    mock_entry = device.get_mock_entry()
+    mock_entry.add_to_hass(hass)
+
+    with patch("broadlink.gendevice", return_value=mock_api), patch.object(
+        hass.config_entries, "async_forward_entry_setup"
+    ), patch.object(hass.config_entries.flow, "async_init"):
+        await hass.config_entries.async_setup(mock_entry.entry_id)
+
+    with patch.object(
+        hass.config_entries, "async_forward_entry_unload", return_value=True
+    ) as mock_forward:
+        assert await hass.config_entries.async_unload(mock_entry.entry_id)
+
+    assert mock_entry.state == ENTRY_STATE_NOT_LOADED
+    assert mock_forward.call_count == 0
+
+
+async def test_device_unload_update_failed(hass):
+    """Test we unload a device that failed the update step."""
+    device = pick_device(1)
+    mock_api = device.get_mock_api()
+    mock_entry = device.get_mock_entry()
+    mock_entry.add_to_hass(hass)
+
+    mock_update_manager = MagicMock()
+    mock_update_manager.coordinator.last_update_success = False
+    mock_update_manager.coordinator.async_refresh = AsyncMock()
```

Comment:
Don't patch details of the integration. We should patch in the periphery, ie the broadlink library.

Reply:
Ok. Done.